Tagged data seems not to work as expected

The following CDDL:

```cddl
start = [ * help ]

help = #6.123(bstr)
```

With the following CBOR (diag notation, hex is `81D87B4100`):

```cbor
[ 123(h'00') ]
```

Does not validate using this library. The Ruby `cddl` gem validates it correctly. The errors are as follow:

```
$ cddl validate --cddl 2.cddl --cbor 2.cbor                                                                                                                                                                                                          
[ERROR] Validation of "5.cbor" failed: error validating group entry associated with rule "help" at cbor location : expected tagged data #6.123("hello"), got Array([Tag(123, Text("hello"))])
```

Edit: I just tested with the main branch and I get the same result.
Edit2: further simplified the CDDL and CBOR.

--- a/src/validator/cbor.rs
+++ b/src/validator/cbor.rs
@@ -418,6 +418,7 @@ impl<'a> CBORValidator<'a> {
                 }
                 ArrayItemToken::Group(group) => cv.visit_group(group)?,
                 ArrayItemToken::Identifier(ident) => cv.visit_identifier(ident)?,
+                ArrayItemToken::TaggedData(tagged_data) => cv.visit_type2(tagged_data)?,
               }
 
               if self.is_multi_type_choice && cv.errors.is_empty() {
@@ -471,6 +472,7 @@ impl<'a> CBORValidator<'a> {
                   }
                   ArrayItemToken::Group(group) => cv.visit_group(group)?,
                   ArrayItemToken::Identifier(ident) => cv.visit_identifier(ident)?,
+                  ArrayItemToken::TaggedData(tagged_data) => cv.visit_type2(tagged_data)?,
                 }
 
                 self.errors.append(&mut cv.errors);
@@ -1073,8 +1075,7 @@ where
           }
           Type2::Array { group, .. } => {
             if let Value::Array(_) = &self.cbor {
-              let entry_counts = entry_counts_from_group(self.cddl, group);
-              self.entry_counts = Some(entry_counts);
+              self.entry_counts = Some(entry_counts_from_group(self.cddl, group));
               self.visit_type2(controller)?;
               self.entry_counts = None;
               return Ok(());
@@ -1728,8 +1729,7 @@ where
             return Ok(());
           }
 
-          let entry_counts = entry_counts_from_group(self.cddl, group);
-          self.entry_counts = Some(entry_counts);
+          self.entry_counts = Some(entry_counts_from_group(self.cddl, group));
           self.visit_group(group)?;
           self.entry_counts = None;
 
@@ -1753,8 +1753,7 @@ where
         Value::Map(m) if self.is_member_key => {
           let current_location = self.cbor_location.clone();
 
-          let entry_counts = entry_counts_from_group(self.cddl, group);
-          self.entry_counts = Some(entry_counts);
+          self.entry_counts = Some(entry_counts_from_group(self.cddl, group));
 
           for (k, v) in m.iter() {
             #[cfg(all(feature = "additional-controls", target_arch = "wasm32"))]
@@ -2030,6 +2029,7 @@ where
           self.errors.append(&mut cv.errors);
           Ok(())
         }
+        Value::Array(_) => self.validate_array_items(&ArrayItemToken::TaggedData(t2)),
         _ => {
           if let Some(tag) = tag {
             self.add_error(format!(
@@ -3865,6 +3865,36 @@ mod tests {
         base16::decode("BA7816BF8F01CFEA414140DE5DAE2223B00361A396177A9CB410FF61F20015AD").unwrap(),
       ),
     ]);
+
+    let cddl = cddl.unwrap();
+
+    let mut cv = CBORValidator::new(&cddl, cbor, None);
+    cv.validate()?;
+
+    Ok(())
+  }
+
+  #[test]
+  fn tagged_data_in_array_validation() -> std::result::Result<(), Box<dyn std::error::Error>> {
+    use ciborium::value::Value;
+
+    let cddl = indoc!(
+      r#"
+        start = [ * help ]
+
+        help = #6.123(bstr)
+      "#
+    );
+
+    let cddl = cddl_from_str(cddl, true).map_err(json::Error::CDDLParsing);
+    if let Err(e) = &cddl {
+      println!("{}", e);
+    }
+
+    let cbor = Value::Array(vec![Value::Tag(
+      123,
+      Box::from(Value::Bytes(base16::decode("00").unwrap())),
+    )]);
 
     let cddl = cddl.unwrap();
 

--- a/src/validator/json.rs
+++ b/src/validator/json.rs
@@ -417,6 +417,7 @@ impl<'a> JSONValidator<'a> {
                 }
                 ArrayItemToken::Group(group) => jv.visit_group(group)?,
                 ArrayItemToken::Identifier(ident) => jv.visit_identifier(ident)?,
+                _ => (),
               }
 
               if self.is_multi_type_choice && jv.errors.is_empty() {
@@ -462,6 +463,7 @@ impl<'a> JSONValidator<'a> {
                 }
                 ArrayItemToken::Group(group) => jv.visit_group(group)?,
                 ArrayItemToken::Identifier(ident) => jv.visit_identifier(ident)?,
+                _ => (),
               }
 
               self.errors.append(&mut jv.errors);
@@ -1169,8 +1171,7 @@ impl<'a, 'b> Visitor<'a, 'b, Error> for JSONValidator<'a> {
         }
         Type2::Array { group, .. } => {
           if let Value::Array(_) = &self.json {
-            let entry_counts = entry_counts_from_group(self.cddl, group);
-            self.entry_counts = Some(entry_counts);
+            self.entry_counts = Some(entry_counts_from_group(self.cddl, group));
             self.visit_type2(controller)?;
             self.entry_counts = None;
             return Ok(());
@@ -1582,8 +1583,7 @@ impl<'a, 'b> Visitor<'a, 'b, Error> for JSONValidator<'a> {
             return Ok(());
           }
 
-          let entry_counts = entry_counts_from_group(self.cddl, group);
-          self.entry_counts = Some(entry_counts);
+          self.entry_counts = Some(entry_counts_from_group(self.cddl, group));
           self.visit_group(group)?;
           self.entry_counts = None;
 

--- a/src/validator/mod.rs
+++ b/src/validator/mod.rs
@@ -1101,6 +1101,7 @@ pub enum ArrayItemToken<'a> {
   Range(&'a Type2<'a>, &'a Type2<'a>, bool),
   Group(&'a Group<'a>),
   Identifier(&'a Identifier<'a>),
+  TaggedData(&'a Type2<'a>),
 }
 
 #[allow(missing_docs)]
@@ -1139,6 +1140,16 @@ impl ArrayItemToken<'_> {
           format!("expected type {} at index {}", ident, idx)
         } else {
           format!("expected type {}", ident)
+        }
+      }
+      ArrayItemToken::TaggedData(tagged_data) => {
+        if let Some(idx) = idx {
+          format!(
+            "expected tagged data tag {:?} at index {}",
+            tagged_data, idx
+          )
+        } else {
+          format!("expected tagged data {:?}", tagged_data)
         }
       }
     }

